File: cli/lib/methods/translated.js
/**
 * Translated (Lara) — professional MT via Translated's Lara Translate API.
 *
 *   Endpoint: official @translated/lara SDK (REST /v2/translate under the
 *             hood; requests are HMAC-signed by the SDK — we never hand-roll
 *             the signature)
 *   Auth:     LARA_ACCESS_KEY_ID + LARA_ACCESS_KEY_SECRET (dashboard →
 *             Settings → API Keys at laratranslate.com)
 *   Codes:    BCP-47; base ISO 639-1 codes are accepted and auto-resolve to
 *             the default regional locale (en → en-US), so we pass our locale
 *             codes through unchanged.
 *   Batch:    the SDK accepts string arrays; we chunk to keep request sizes
 *             polite and failures isolated.
 *
 * ModernMT note: Translated is sunsetting ModernMT into Lara (API-compatible
 * migration path until end of 2026); this adapter targets Lara only.
 */

import { TranslationMethod } from './base.js';
import { getEnvOrFileVar } from '../api-key.js';
import { DEFAULT_METHOD_CONCURRENCY } from '../config.js';
import { extractContentBody } from './content-separator.js';
import { output } from '../output.js';
import { pMap } from '../concurrent.js';

const TRANSLATED_MAX_BATCH = 50;

class TranslatedMethod extends TranslationMethod {
  constructor(options = {}) {
    super('translated', options);
    this._client = options.laraClient || null; // injectable for tests
  }

  // ── Credential / client resolution ─────────────────────────────

  /**
   * Resolve the Lara access-key pair from options, env vars, or .env files.
   * @param {object} options - Caller-provided options
   * @returns {{id: string, secret: string}|null}
   */
  _resolveCredentials(options = {}) {
    const id = options.laraAccessKeyId
      || getEnvOrFileVar('LARA_ACCESS_KEY_ID')
      || getEnvOrFileVar('LARA_ACCESS_KEY_ID', options.cwd);
    const secret = options.laraAccessKeySecret
      || getEnvOrFileVar('LARA_ACCESS_KEY_SECRET')
      || getEnvOrFileVar('LARA_ACCESS_KEY_SECRET', options.cwd);
    return id && secret ? { id, secret } : null;
  }

  /**
   * Lazily construct the official Lara SDK client. The SDK owns request
   * signing (HMAC over the request envelope), retries, and endpoint choice.
   * @param {object} options
   * @returns {Promise<object|null>} Translator instance or null
   */
  async _getClient(options = {}) {
    if (this._client) return this._client;
    const creds = this._resolveCredentials(options);
    if (!creds) return null;
    let lara;
    try {
      lara = await import('@translated/lara');
    } catch {
      output.error(
        'Translated (Lara): the @translated/lara SDK is not installed — '
        + 'run `npm install` in the champollion package.',
      );
      return null;
    }
    const { Credentials, Translator } = lara;
    this._client = new Translator(new Credentials(creds.id, creds.secret));
    return this._client;
  }

  // ── Translation ────────────────────────────────────────────────

  /**
   * Translate one chunk of texts. Returns the translations array (in input
   * order) or null on failure — callers treat null as a skipped batch.
   */
  async _translateBatch(client, sourceTexts, sourceLocale, targetLocale, batchNum) {
    try {
      const result = await client.translate(sourceTexts, sourceLocale, targetLocale);
      const translations = Array.isArray(result?.translation)
        ? result.translation
        : [result?.translation];
      if (translations.length !== sourceTexts.length) {
        output.error(
          `Translated batch ${batchNum}: response length mismatch `
          + `(sent ${sourceTexts.length}, got ${translations.length})`,
        );
        return null;
      }
      return translations;
    } catch (err) {
      output.error(`Translated batch ${batchNum}: ${err?.message || err}`);
      return null;
    }
  }

  /**
   * Translate a batch of key-value pairs via the Lara API.
   *
   * @param {string[]} keys - Flat dot-notation keys to translate
   * @param {object} sourceFlat - Full flattened source locale
   * @param {import('../types.js').PairConfig} pairConfig - Pair config
   * @param {object} options - { laraAccessKeyId, laraAccessKeySecret, batchSize }
   * @returns {object|null} Map of key → translated value, or null
   */
  async translate(keys, sourceFlat, pairConfig, options) {
    const client = await this._getClient(options);
    if (!client) {
      output.warn(
        'Translated (Lara): no credentials (LARA_ACCESS_KEY_ID / '
        + 'LARA_ACCESS_KEY_SECRET) — skipping.',
      );
      return null;
    }

    const sourceLocale = pairConfig.source || 'en';
    const targetLocale = pairConfig.target;
    const maxSegments = pairConfig.batchSize || options.batchSize || TRANSLATED_MAX_BATCH;
    const allTranslated = {};

    const batchChunks = [];
    for (let i = 0; i < keys.length; i += maxSegments) {
      batchChunks.push(keys.slice(i, i + maxSegments));
    }

    await pMap(batchChunks, async (chunk, idx) => {
      const orderedKeys = [];
      const sourceTexts = [];
      for (const key of chunk) {
        const val = sourceFlat[key];
        if (val && typeof val === 'string') {
          orderedKeys.push(key);
          sourceTexts.push(val);
        }
      }
      if (sourceTexts.length === 0) return;

      const translations = await this._translateBatch(
        client, sourceTexts, sourceLocale, targetLocale, idx + 1,
      );
      if (!translations) return;
      for (let i = 0; i < orderedKeys.length; i++) {
        if (typeof translations[i] === 'string') {
          allTranslated[orderedKeys[i]] = translations[i];
        }
      }
    }, { concurrency: DEFAULT_METHOD_CONCURRENCY });

    return Object.keys(allTranslated).length > 0 ? allTranslated : null;
  }

  /**
   * Translate freeform Markdown content via Lara.
   * @param {string} prompt - Content prompt (body extracted before sending)
   * @param {import('../types.js').PairConfig} pairConfig
   * @param {object} options
   * @returns {Promise<string|null>}
   */
  async translateContent(prompt, pairConfig, options) {
    const client = await this._getClient(options);
    if (!client) return null;
    const body = extractContentBody(prompt);
    if (!body.trim()) return null;
    const translations = await this._translateBatch(
      client, [body], pairConfig.source || 'en', pairConfig.target, 'content',
    );
    return translations ? translations[0] : null;
  }

  // ── Metadata ───────────────────────────────────────────────────

  checkReadiness(context) {
    const creds = this._resolveCredentials(context || {});
    if (!creds) {
      return {
        ready: false,
        reason: 'No Lara credentials (LARA_ACCESS_KEY_ID + LARA_ACCESS_KEY_SECRET). '
          + 'Create an API key at laratranslate.com → Settings → API Keys.',
      };
    }
    return { ready: true };
  }

  estimateCost(_keyCount) {
    // Proprietary per-character pricing behind Translated's plans; never
    // invent a number (Project rule: never invent pricing).
    return {
      estimatedCost: null,
      currency: 'USD',
      source: 'translated-lara-proprietary',
      note: 'Lara pricing is plan-based; per-character cost not published.',
    };
  }

  getQualityTier() {
    return 'standard';
  }

  getProvenance() {
    return {
      resources: [
        {
          name: 'Translated Lara Translate API',
          license: 'Proprietary (Translated ToS)',
          type: 'api',
        },
      ],
      commercialReady: true,
      flags: [],
    };
  }

  getSetupHelp() {
    return [
      '  Translated (Lara) — professional MT API, 200+ languages:',
      '   1. Sign up: https://laratranslate.com',
      '   2. Settings → API Keys → create an access key',
      '   3. export LARA_ACCESS_KEY_ID=...  LARA_ACCESS_KEY_SECRET=...',
      '      (or add both to .env.local)',
    ];
  }
}

export { TranslatedMethod, TRANSLATED_MAX_BATCH };

File: arena/mt_eval_harness/methods/translated.py
"""
TranslatedMethod — Translated (Lara) professional MT adapter.

Lara is Translated's production MT platform (the ModernMT lineage; ModernMT
sunsets into Lara at the end of 2026). Requests are HMAC-signed, so this
adapter goes through the official ``lara-sdk`` package rather than
hand-rolling the signature — the SDK owns auth, endpoint, and retries.

    SDK:       pip install lara-sdk   (lazy-imported; clear error if absent)
    Auth:      LARA_ACCESS_KEY_ID + LARA_ACCESS_KEY_SECRET
               (laratranslate.com → Settings → API Keys)
    Batch:     the SDK translate() accepts a list of strings
    Locale:    BCP-47; base ISO 639-1 codes auto-resolve to the default
               regional locale (en → en-US), so card codes pass through
    Cost:      proprietary, plan-based; never invent pricing
"""

from __future__ import annotations

import asyncio

from mt_eval_harness.methods.base_http_mt import (
    HttpMTMethod,
    MTConfigError,
    _env_first,
)


class TranslatedMethod(HttpMTMethod):
    """Translated (Lara) MT API system-under-test."""

    name = "translated"
    MAX_BATCH = 50

    method_id = "translated-lara"
    paradigm = "neural-nmt"  # method_class="api" inherited
    PROVIDER_CODE_FIELD = "iso639_1"
    author = "Translated"
    description = (
        "Translated (Lara) — professional MT platform (200+ languages; "
        "ModernMT lineage), adaptive translation with the official SDK."
    )
    homepage = "https://laratranslate.com"
    license = "Proprietary (Translated ToS)"
    commercial_ready = True
    cost_note = "Proprietary; plan-based pricing (per-character cost not published)"

    # --- Locale mapping ---------------------------------------------------

    @staticmethod
    def _map_locale(code: str) -> str:
        """Lara accepts base ISO 639-1 codes (defaults the regional locale)."""
        return code

    # --- Credentials ------------------------------------------------------

    def _resolve_credentials(self) -> dict:
        key_id = self.options.get("lara_access_key_id") or _env_first(
            "LARA_ACCESS_KEY_ID"
        )
        key_secret = self.options.get("lara_access_key_secret") or _env_first(
            "LARA_ACCESS_KEY_SECRET"
        )
        if not key_id or not key_secret:
            raise MTConfigError(
                "Translated (Lara): missing credentials. Set LARA_ACCESS_KEY_ID "
                "and LARA_ACCESS_KEY_SECRET (laratranslate.com → Settings → "
                "API Keys)."
            )
        return {"key_id": key_id, "key_secret": key_secret}

    # --- SDK client (factored out so tests can inject a fake) --------------

    _translator = None  # cached SDK Translator instance

    def _get_translator(self, creds: dict):
        if self._translator is not None:
            return self._translator
        try:
            from lara import Credentials, Translator  # lazy: optional dep
        except ImportError as e:
            raise MTConfigError(
                "Translated (Lara): the 'lara-sdk' package is not installed — "
                "run `pip install lara-sdk`."
            ) from e
        self._translator = Translator(
            Credentials(creds["key_id"], creds["key_secret"])
        )
        return self._translator

    @staticmethod
    def _parse_result(result, expected: int) -> list[str]:
        """Normalize the SDK's TextResult into an ordered list of strings."""
        translation = getattr(result, "translation", None)
        if translation is None and isinstance(result, dict):
            translation = result.get("translation")
        if isinstance(translation, str):
            translation = [translation]
        if not translation or len(translation) != expected:
            got = len(translation) if translation else 0
            raise ValueError(
                f"Translated: response length mismatch (expected {expected}, got {got})"
            )
        return [str(t) for t in translation]

    # --- Backend ----------------------------------------------------------

    async def _translate_texts(
        self,
        texts: list[str],
        src: str,
        tgt: str,
        creds: dict,
    ) -> list[str]:
        translator = self._get_translator(creds)
        # The SDK is synchronous; run it off the event loop.
        result = await asyncio.to_thread(
            translator.translate,
            list(texts),
            self._map_locale(src),
            self._map_locale(tgt),
        )
        return self._parse_result(result, len(texts))
